### PR TITLE
Debug jumps in etendue for spectrometer concave crystals

### PR DESCRIPTION
**Pull request by dvezinet**
Main changes:
-----------------
Two minor bugs identified in the routines used to project an aperture (slit) on a concave crystal.
Concavity comes with numerical instability when the root-finding process comes close to the singular point that is the center of curvature.

Tofu implements a robust root-finding routine + interpolates reflected polygons to take out singular points.

**Bug 1:** the interpolation process was missing some corners of the polygons in some particular cases.

**Bug 2:** the root-finding algo wasn't robust enough for some particular pixels / configurations

Both bugs have been addressed in this PR, see issue #747 for examples

Several plotting / debugging sections have been added to the code to facilitate any later debugging in case another problem arises. They are commented out and left as such for later use.

Issues:
-------
* Fixes, in devel, issue #747

### CODE REVIEW

File: tofu/data/_class8_equivalent_apertures.py [hunk 1]
```diff
@@ -237,10 +255,9 @@ def equivalent_apertures(
 
             elif ln[ii] < nmax:
                 ndif = nmax - ln[ii]
-                irand = np.random.random(ndif)
+                irand = np.random.randint(1, 9, ndif)/10.
                 irand = irand + np.random.randint(0, ln[ii]-1, ndif)
                 imax = np.sort(np.r_[np.arange(0, ln[ii]), irand])
-                imax = np.linspace(0, ln[ii]-1, nmax)
```
Comment (dvezinet) [inline tofu/data/_class8_equivalent_apertures.py:243]: First bug was these 4 lines

File: tofu/data/_class5_reflections_pts2pt.py [hunk 2]
```diff
@@ -247,21 +247,35 @@ def pts2pt(
...
+                    thr = abs(line[0] - line[-1]) * 0.05
+                    std = np.std(eq[ind] - line)
+                    if std > thr:
+                        continue
+                    roots = np.r_[-bb / aa]
+                    # roots = np.r_[
+                    #     kk[i0]
+                    #     - eq[i0]*(kk[i0+1] - kk[i0])
+                    #     / (eq[i0+1] - eq[i0])
+                    # ]
```
Comment (dvezinet) [inline tofu/data/_class5_reflections_pts2pt.py:278]: Second bug: more robust root-finding

**Comment by cjperks7** [comment]
Just flipping plotting data through `diag.plot_diagnostic()` things look good except for `dlamb` and `res`. This is more so related to what's stated in Issue #748 than what's addressed here. Attached is a plot of dlamb from `diag.plot_diagnostic()`. Also attached is repeating the comparison plot given in Issue #748 to illustrate that the factor is still about off the same amount, but the shape changed from what you did here
Screenshot 2023-04-27 at 5 57 33 PM
Screenshot 2023-04-27 at 6 00 20 PM

**Review by cjperks7 — APPROVED** [review]

**Comment by Didou09** [comment]
Well spotted (dlamb), as you mention, it's more related to issue #748 , will try to address that tomorrow morning

**Comment by cjperks7** [comment]
Definitely changed some things up with the magnitude of `dlamb`. The prefactor to rescale it to just being the spacing in the wavelength mesh is now ~3.5 (previously ~4.3). 

Much less noisy though. The relative difference between the two meshes shown is now 1e-6%. So any difference in the first attached plot is really the subtraction I did for visualization.

The additional computational time isn't terrible. Might have slowed down "build diagnostic" section in tofu_sparc by say 50%-ish so not the end of the world.

Screenshot 2023-04-28 at 1 55 10 PM

**Pull request merged**